README

			AppPlusShell Widget
----------------------------------------------------------------------


From the Xt FAQ:

----------------------------------------------------------------------
4.  How do I use a different visual than the default?
----------------------------------------------------------------------

This requires a more complicated answer than it should.  A window has
three things that are visual specific -- the visual, colormap and
border pixmap.  All widgets have their own Colormap and BorderPixmap
resource; only shell widgets have Visual resources (another questions
deals with why shells have a Visual).  The default value of these
resources is CopyFromParent which does exactly what it says.  In the
shell widget CopyFromParent gets evalulated as DefaultVisualOfScreen
and DefaultColormapOfScreen.  When any one of the three resources is
not properly set, a BadMatch error occurs when the window is
created.  They are not properly set because each of the values depends
on the visual being used.

How to get this to work?  There are two parts to the answer.  The
first is if you want an application to start with a particular visual
and the second is if you want a particular shell within an application
to start with a different visual.  The second is actually easier
because the basic information you need is available.  The first is a
little harder because you'll need to initialize much of the toolkit
yourself in order to determine the needed information.

----------------------------------------------------------------------


	This doesn't seem like the correct solution to me.  First, its
not reusable, and second it's not user configurable thru resources.  
This was the inspiration behind the AppPlusShell widget.  

	The AppPlusShell adds the following to the application shell:

1) Visual/Colormap control thru resources
2) EditRes built in (can be turned on/off)
3) Catches the WM_DELETE_WINDOW

	It does this thru the following added resources:

For 1:
XtNvisualClass
XtNusePrivateColormap
XtNvisualID
XtNapplicationDepth

For 2:
XtNallowEditRes

For 3:
XtNsaveYourselfCallback



	The algorithm is as follows, in this order:

1) if XtNvisualID is set try to use the visual with this id, if
it exists.

2) if XtNapplicationDepth is set try to find a visual with 
that depth, otherwise set it to DefaultDepth().  This allows 
users to have:

*applicationDepth: 24

which will be used if it exists.  XtNapplicationDepth is needed
instead of XtNdepth, since, IMO, most users will want to do

*depth: 24 

which would screw up for any subwidgets if there were no visual
with a depth of 24.  Note that the depth gets correctly set in the 
initialize too.

3) if XtNvisualClass is set then find the visual with the that
class and the greatest depth.

4) if all else fails use the default visual/default depth.

5a) if he visualID == default visual id, and XtNusePrivateColormap
is FALSE, then use the default colormap, otherwise, create an
appropriate one.

5b) If the XCC library is in use, it would call the XCC lib
at this point instead.

	Plus, there is now a file called SubPlusS.c that containts
functions for creating popup/menu shells with the correct
visuals/colormaps/depths.


	The only other thing of interest is if you define MOTIF
the code uses some Motif functions instead of their X/Xt
equivalents, but otherwise should work with pretty much any
Xt widget set.

	I use a version of this code in our product, and so it has
been used on just about every major U*x platform with success.

	Let me know what you think...(Otherwise I'll stop writing this
stuff...Well, after I finish this next one, then that's it..., Yeah,
that's it...)

	John L. Cwikla
	cwikla@wri.com


CHANGES

/* $Id: CHANGES,v 1.22 1996/04/21 16:58:53 cwikla Exp $ */

Added SubPlusS.[ch] for popup/menu shells.
Got rid of extraneous variables

Added <Xm/Xm.h> for -DMOTIF so that XmREVISON would exist!


Fixed check if there is a good default depth and the visualClass
is not set, it would try to match a visual class of -1.


Fixed SubPlus.c XtPlusCreatePopup[Shell/Menu] where I was
accessing the wrong indexes of the incoming Args.
Thanks to Kensaku MORI (mori@eagle.toriwaki.nuie.nagoya-u.ac.jp)

Mon Jan 30 11:38:47 CST 1995

Added XCC library.

Mon Jan 30 13:32:10 CST 1995


Fixed screwed up direct color/true color

Mon Jan 30 14:17:55 CST 1995


Fixed screwed up standard colormap atom

Mon Jan 30 14:49:45 CST 1995

Fix bit count for _initDirect

Tue Jan 31 17:02:28 CST 1995

Fix _atom in _lookForStdCmap to ColorAtomList/GrayAtomList


Wed Feb  1 18:55:32 CST 1995

Moved the XCC to the class part where it belongs.
Added some #ifdef HAS_XCC

Fri Feb  3 12:07:32 CST 1995

Check for red_max/green_max/blue_max to avoid divide by 0;


Tue Feb  7 11:14:45 CST 1995


Fix from Anre Olav Morken (Arne.Morken@ii.uib.no)
variable "cubes" was not initialized and may be doing bogus
frees.

Tue Feb  7 13:23:28 CST 1995

Added XCCGetColormap function
made converters static



Sun Feb 12 12:32:36 CST 1995


Changed example program
Set background_pixel, and border_color to 0 in AppPlusS.c realize proc.
Did same in SubPlus.c

Thu Feb 16 21:59:48 CST 1995

Fixed dinc in _initGray routine.

Tue Mar 14 12:11:19 CST 1995

Added New public routines:

extern XCC XCCMonoCreate(Display *_dpy, Visual *_visual, Colormap _colormap);
extern Display *XCCGetDisplay(XCC _xcc);
extern int XCCQueryColor(XCC _xcc, XColor *_color);


Thu Apr  6 12:56:55 CDT 1995

Fix/Hack for improper RGB_GRAY_MAP's that use/don't use green/blue max

Wed Apr 12 21:39:22 CDT 1995


Fixed numColors for standard colormaps
New demo program

Thu Apr 20 23:04:54 CDT 1995

Made Mono screens always use MODE_BW

Sat May 13 22:00:52 CDT 1995

From Daniel Fu (daniel@cs.uchicago.edu)

void  XCCFree(XCC _xcc)
ADDED:  XFree(_xcc->visualInfo);

void _initGray(XCC _xcc)
CHANGED:  _xcc->stdCmap.red_max = _xcc->numColors-1;

Mon May 15 12:10:46 CDT 1995


Fixed typo



Tue May 16 14:13:50 CDT 1995


Finally!!!! DirectColor is now correct!!!! I was
not using the CLUT like I should have been (what was
I thinking?)


Fri Aug  4 12:45:21 CDT 1995

Added XCCQueryColor/XCCQueryColors().  Now does local querying
of colors if possible.

Wed Aug  9 23:06:56 CDT 1995


Fixed XCCQueryColors() which was still calling XQueryColor()

Sat Aug 12 13:31:14 CDT 1995

Fixed XCCQueryColors not increment tc pointer.

Fri Oct 27 14:18:19 CDT 1995

Fixed a problem where a standard colormap could exist, the visual id
could be right, but a colormap already exists that is NOT the default.

Wed Nov 29 16:19:16 CST 1995

Fixed a == NULL to != NULL, that was keeping the AppPlusS from finding
the deepest depth visual.


Fri Apr 19 10:57:45 CDT 1996

Bugs fixed from bsletten@autometric.com reports:

Now send passed in args for SubPlusS.c to both MenuShell and RowColumn.
This was making tear-off-menus not work.

Made the XCC not do a queryColors malloc if the depth > 12. This could happen
apparantly on a TrueColor/DirectColor visual which is bad.

Fri Apr 19 11:56:09 CDT 1996

Arg. Real fix for SubPlusS.c...

Sun Apr 21 11:52:35 CDT 1996

Subtracted 1 for number of args when concatanating args passes in.

AppPlusS.c
/* $Id: AppPlusS.c,v 1.9 1995/11/29 22:22:56 cwikla Exp $ */
/*
 * Copyright 1994,1995 John L. Cwikla
 *
 * Permission to use, copy, modify, distribute, and sell this software
 * and its documentation for any purpose is hereby granted without fee,
 * provided that the above copyright notice appears in all copies and that
 * both that copyright notice and this permission notice appear in
 * supporting documentation, and that the name of John L. Cwikla or
 * Wolfram Research, Inc not be used in advertising or publicity
 * pertaining to distribution of the software without specific, written
 * prior permission.  John L. Cwikla and Wolfram Research, Inc make no
 * representations about the suitability of this software for any
 * purpose.  It is provided "as is" without express or implied warranty.
 *
 * John L. Cwikla and Wolfram Research, Inc disclaim all warranties with
 * regard to this software, including all implied warranties of
 * merchantability and fitness, in no event shall John L. Cwikla or
 * Wolfram Research, Inc be liable for any special, indirect or
 * consequential damages or any damages whatsoever resulting from loss of
 * use, data or profits, whether in an action of contract, negligence or
 * other tortious action, arising out of or in connection with the use or
 * performance of this software.
 *
 * Author:
 *  John L. Cwikla
 *  X Programmer
 *  Wolfram Research Inc.
 *
 *  cwikla@wri.com
*/

#include <X11/IntrinsicP.h>
#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <X11/Xmu/Editres.h>

#if HAS_XCC
#include "XCC.h"
#endif /* HAS_XCC */

#include "AppPlusSP.h"

#include <string.h>
#include <ctype.h>


#if MOTIF 
#include <Xm/Xm.h>
#if (XmREVISION >= 2)
#include <Xm/Protocols.h>
#endif
#endif /* MOTIF */

#if NeedFunctionProtoTypes
static XtProc classInitialize(void);
static XtInitProc initialize(AppPlusShellWidget _request, AppPlusShellWidget _new, String *_args, Cardinal *_numArgs);
static XtRealizeProc realize(AppPlusShellWidget _apsw, XtValueMask *_xvm, XSetWindowAttributes *_xswa);
static XtWidgetProc destroy(AppPlusShellWidget _apsw);
static void WMProtoAction(AppPlusShellWidget _apsw, XEvent *_event, String *_parms, Cardinal *_numParms);
static Boolean cvtStringToVisualClass(Display *_display, XrmValuePtr _args,
	Cardinal *_numArgs, XrmValuePtr _from, XrmValuePtr _to, XtPointer *_data);
static Boolean cvtStringToVisualID(Display *_display, XrmValuePtr _args,
	Cardinal *_numArgs, XrmValuePtr _from, XrmValuePtr _to, XtPointer *_data);
#else
static XtProc classInitialize();
static XtInitProc initialize();
static XtRealizeProc realize();
static XtWidgetProc destroy();
static void WMProtoAction();
static Boolean cvtStringToVisualClass();
static Boolean cvtStringToVisualID();
#endif /* NeedFunctionProtoTypes */

#define CORE(a) ((a)->core)
#define APPS(a) ((a)->app_plus_shell)
#define SHELL(a) ((a)->shell)
#define ACLASS(a) ((a)->app_plus_shell_class)
#define VID(a) ((a)->app_plus_shell.visualID)

#if MOTIF & (XmREVISION >= 2)
#define XtCharStringToAtom(_w, _string) XmInternAtom(XtDisplay(_w), _string, FALSE)
#else
#define XtCharStringToAtom(_w, _string) XInternAtom(XtDisplay(_w), _string, FALSE)
#endif /* MOTIF & (XmREVISION >= 2) */

#define DEFAULT -1

#define TheOffset(field) XtOffset(AppPlusShellWidget, app_plus_shell.field)
static XtResource resources[] =
{
	{XtNdata, XtCData, XtRPointer, sizeof(XtPointer),
		TheOffset(data), XtRImmediate, (XtPointer)NULL},
	{XtNsaveYourselfCallback, XtCCallback, XtRCallback, sizeof(XtCallbackList),
		TheOffset(saveYourselfCallback), XtRImmediate, (XtPointer)NULL},
	{XtNvisualClass, XtCVisualClass, XtRVisualClass, sizeof(int),
		TheOffset(visualClass), XtRImmediate, (XtPointer)DEFAULT},
	{XtNusePrivateColormap, XtCUsePrivateColormap, XtRBoolean, sizeof(Boolean),
		TheOffset(usePrivateColormap), XtRImmediate, (XtPointer)FALSE},
	{XtNvisualID, XtCVisualID, XtRVisualID, sizeof(VisualID),
		TheOffset(visualID), XtRImmediate, (XtPointer)0},
	{XtNallowEditRes, XtCAllowEditRes, XtRBoolean, sizeof(Boolean),
		TheOffset(allowEditRes), XtRImmediate, (XtPointer)FALSE},
	{XtNapplicationDepth, XtCApplicationDepth, XtRInt, sizeof(int),
		TheOffset(applicationDepth), XtRImmediate, (XtPointer)DEFAULT},
#if HAS_XCC
	{XtNuseStandardColormaps, XtCUseStandardColormaps, XtRBoolean, sizeof(Boolean),
		TheOffset(useStandardColormaps), XtRImmediate, (XtPointer)TRUE},
	{XtNstandardColormap, XtCStandardColormap, XtRAtom, sizeof(Atom),
		TheOffset(standardColormap), XtRImmediate, (XtPointer)0},
#endif /* HAS_XCC */
};
#undef TheOffset

static char defaultTranslations[] =
"<ClientMessage> WM_PROTOCOLS: WMProtoAction()";
 
static XtActionsRec actions[] =
{
	{"WMProtoAction", (XtActionProc)WMProtoAction},
};

AppPlusShellClassRec appPlusShellClassRec = 
{
	{	/* Core */
		(WidgetClass)&applicationShellClassRec,		 /* superclass */
		"AppPlusShell",								/* class_name */
		sizeof(AppPlusShellRec),					 /* widget_size */
		(XtProc)classInitialize,				 /* class_initialize */
		(XtWidgetClassProc)NULL,	 /* class_part_initialize */
		FALSE,						/* class_init */
		(XtInitProc)initialize,		 /* initialize */
		NULL,						 /* initialize_hook */
		(XtRealizeProc)realize,		 /* realize */
		actions,					/* actions */
		XtNumber(actions),			/* num_actions */
		resources,					 /* resources */
		XtNumber(resources),		 /* num_resouces */
		NULLQUARK,					/* xrm_class */
		FALSE,						 /* compress_motion */
		FALSE,	 					/* compress_exposure */
		FALSE,					 /* compress_enter_leave */
		FALSE,					 /* visibility_interest */
		(XtWidgetProc)destroy,	 /* destroy */
		XtInheritResize,		/* resize */
		XtInheritExpose,		/* expose */
		NULL,				 /* setvalues */
		NULL,				 /* setvalues_hook */
		XtInheritSetValuesAlmost, /* setvalues_almost */
		NULL,						 /* get_values_hook */
		(XtAcceptFocusProc)XtInheritAcceptFocus,	 /* accept_focus */
		XtVersion,							/* version */
		NULL,								 /* callback_private */
		defaultTranslations,				 /* tm_translations */
		XtInheritQueryGeometry,				 /* query_geometry */
		XtInheritDisplayAccelerator,		/* display_accelerator */
		NULL,								 /* extension */
	},
	{	/* Composite */
		XtInheritGeometryManager,
		XtInheritChangeManaged,
		XtInheritInsertChild,
		XtInheritDeleteChild,
		(XtPointer)NULL,
	},
	{	/* Shell */
		0,
	},
	{	/* WindowManagerShell */
		0,
	},
	{	/* VendorShell */
		(XtPointer)NULL,
	},
	{	/* TopLevelShell */
		(XtPointer)NULL,
	},
	{	/* ApplicationShell */
		(XtPointer)NULL,
	},
	{ 	/* AppPlusShell */
		(Colormap)0,
		(Visual *)NULL,
		0,	
		0,
#if HAS_XCC
		NULL,
#endif /* HAS_XCC */
	},
};

WidgetClass appPlusShellWidgetClass = (WidgetClass)&appPlusShellClassRec;

#if DEBUGGING
static char *VisualClassToName(int class)
{
	switch(class)
	{
		case StaticGray:
			return "StaticGray";
		case StaticColor:
			return "StaticColor";
		case GrayScale:
			return "GrayScale";
		case PseudoColor:
			return "PseudoColor";
		case TrueColor:
			return "TrueColor";
		case DirectColor:
			return "DirectColor";
		default:
			return "Unknown";
	}
}
#endif /* DEBUGGING */

#if NeedFunctionProtoTypes
static XtProc classInitialize(void)
#else
static XtProc classInitialize()
#endif /* NeedFunctionProtoTypes */
{
	XtSetTypeConverter(XtRString, XtRVisualClass, cvtStringToVisualClass,
		(XtConvertArgList)NULL, 0,
		XtCacheAll, (XtDestructor)NULL);

	XtSetTypeConverter(XtRString, XtRVisualID, cvtStringToVisualID,
		(XtConvertArgList)NULL, 0,
		XtCacheAll, (XtDestructor)NULL);
}

#if NeedFunctionProtoTypes
static void useDefaults(AppPlusShellWidget _apsw)
#else
static void useDefaults(_apsw)
AppPlusShellWidget _apsw;
#endif /* NeedFunctionProtoTypes */
{
	CORE(_apsw).depth = DefaultDepth(XtDisplay(_apsw), DefaultScreen(XtDisplay(_apsw)));
	SHELL(_apsw).visual = DefaultVisual(XtDisplay(_apsw), DefaultScreen(XtDisplay(_apsw)));
	APPS(_apsw).visualClass = SHELL(_apsw).visual->class;
}

#if NeedFunctionProtoTypes
static Boolean validVisualID(AppPlusShellWidget _apsw, XVisualInfo *_visualInfo)
#else
static Boolean validVisualID(_apsw, _visualInfo)
AppPlusShellWidget _apsw;
XVisualInfo *_visualInfo;
#endif /* NeedFunctionProtoTypes */
{
	XVisualInfo *vinfos, vtemp;
	int vitems;

	if (VID(_apsw) == 0)
		return FALSE;

	vtemp.visualid = VID(_apsw);

	vinfos = XGetVisualInfo(XtDisplay(_apsw), VisualIDMask, &vtemp, &vitems);

	if (vinfos != NULL)
	{
/*
** There better ONLY BE ONE!
*/
		_visualInfo->visual = vinfos[0].visual;
		_visualInfo->depth = vinfos[0].depth;
		_visualInfo->class = vinfos[0].class;

		XFree((char *)vinfos);
		return TRUE;
	}
	else
		return FALSE;
}

#if NeedFunctionProtoTypes
static XtInitProc initialize(AppPlusShellWidget _request, AppPlusShellWidget _new, String *_args, Cardinal *_numArgs)
#else
static XtInitProc initialize(_request, _new, _args, _numArgs)
AppPlusShellWidget _request;
AppPlusShellWidget _new;
String *_args;
Cardinal *_numArgs;
#endif /* NeedFunctionProtoTypes */
{
	Display *display;
	int screenNumber;
	Visual *defVisual;
	XVisualInfo visualInfo;
	AppPlusShellWidgetClass aswc;

	if (APPS(_new).allowEditRes)
		XtAddEventHandler((Widget)_new, (EventMask)0, TRUE, _XEditResCheckMessages , NULL);

	aswc = (AppPlusShellWidgetClass)appPlusShellWidgetClass;

	if (ACLASS(aswc).colormap == 0)
	{
#if DEBUGGING
		printf("Creating Visual, colormap, etc...\n");
#endif /* DEBUGGING */

		display = XtDisplay(_new);
		screenNumber = DefaultScreen(display);
		defVisual = DefaultVisual(display, screenNumber);

/*
** Highest priority is given to a specified visual id.
*/
		if (validVisualID(_new, &visualInfo))
		{
				SHELL(_new).visual = visualInfo.visual;
				CORE(_new).depth = visualInfo.depth;
				APPS(_new).visualClass = visualInfo.class;
		}
		else
		if ( (APPS(_new).applicationDepth == DEFAULT) && (APPS(_new).visualClass == DEFAULT))
		{
			useDefaults(_new);
		}
		else
		{
			if (APPS(_new).applicationDepth == DEFAULT)
				CORE(_new).depth = DefaultDepth(display, screenNumber);
			else
				CORE(_new).depth = APPS(_new).applicationDepth;

			if (APPS(_new).visualClass == DEFAULT)
				APPS(_new).visualClass = defVisual->class;
/*
** Try for match.
*/
			if (XMatchVisualInfo(display, screenNumber, CORE(_new).depth, 
				APPS(_new).visualClass, &visualInfo) != 0)
			{
				SHELL(_new).visual = visualInfo.visual;
				CORE(_new).depth = visualInfo.depth;
				APPS(_new).visualClass = visualInfo.class;
			}
			else
			{
				XVisualInfo visTemplate;
				XVisualInfo *visReturn;
				int numVis, n;

				n = 0;
				visReturn = (XVisualInfo *)NULL;

/*
** Next see if we can find a depth that they have set.
*/

				if (APPS(_new).applicationDepth != DEFAULT)
				{
					visTemplate.depth = CORE(_new).depth;
					visReturn = XGetVisualInfo(display, VisualDepthMask, &visTemplate, &numVis);

/*
** If numVis is greater than 1 you may want to have it pick your favorite visual here. I
** don't think this is really necessary though, since the user still has control with
** the XtNvisualClass to get the one he/she really wants.
*/
				}

				if (visReturn == (XVisualInfo *)NULL)
				{
/*
** Nope.  Now find the best depth at this visual class
*/
					visTemplate.class = APPS(_new).visualClass;
					visReturn = XGetVisualInfo(display, VisualClassMask, &visTemplate, &numVis);

					if (visReturn != (XVisualInfo *)NULL)
					{
						int i;
						n = 0;
						for (i=1;i<numVis;i++)
							if (visReturn[i].depth > visReturn[n].depth)
								n = i;
					}
				}

/*
** Nope. Use defaults.
*/
				if (visReturn == (XVisualInfo *)NULL)
					useDefaults(_new);
				else
				{
					SHELL(_new).visual =	visReturn[n].visual;
					CORE(_new).depth = visReturn[n].depth;
					APPS(_new).visualClass = visReturn[n].class;
					XtFree((char *)visReturn);
				}
			}
		}

#if HAS_XCC
		ACLASS(aswc).xcc = XCCCreate(display, SHELL(_new).visual, APPS(_new).usePrivateColormap, 
							APPS(_new).useStandardColormaps, APPS(_new).standardColormap, &CORE(_new).colormap);
#else


		if ( (SHELL(_new).visual->visualid == defVisual->visualid) && !APPS(_new).usePrivateColormap)
		{
			CORE(_new).colormap =   DefaultColormap(display, screenNumber);
#if DEBUGGING
			printf("Using default visual, depth, and colormap\n");
#endif /* DEBUGGING */
		}
		else
		{
			CORE(_new).colormap = XCreateColormap(display, RootWindow(display, screenNumber), 
				SHELL(_new).visual, AllocNone);
#if DEBUGGING
			printf("Created colormap\n");
#endif /* DEBUGGING */
		}
#endif /* HAS_XCC */


		ACLASS(aswc).visual = SHELL(_new).visual;
		ACLASS(aswc).colormap = CORE(_new).colormap;
		ACLASS(aswc).visualClass = APPS(_new).visualClass;
		ACLASS(aswc).applicationDepth = CORE(_new).depth;
	}
	else
	{
#if DEBUGGING
		printf("Using already created visual, colormap, etc...\n");
#endif /* DEBUGGING */
		SHELL(_new).visual = ACLASS(aswc).visual;
		CORE(_new).colormap = ACLASS(aswc).colormap;
		APPS(_new).visualClass = ACLASS(aswc).visualClass;
		CORE(_new).depth = ACLASS(aswc).applicationDepth;
	}


	APPS(_new).numberProtos = 0;
	APPS(_new).saveYourself[(APPS(_new).numberProtos)++] = XtCharStringToAtom((Widget)_new, "WM_DELETE_WINDOW");
}

#if NeedFunctionProtoTypes
static XtRealizeProc realize(AppPlusShellWidget _apsw, XtValueMask *_xvm, XSetWindowAttributes *_xswa)
#else
static XtRealizeProc realize(_apsw, _xvm, _xswa)
AppPlusShellWidget _apsw;
XtValueMask *_xvm;
XSetWindowAttributes *_xswa;
#endif /* NeedFunctionProtoTypes */
{
	_xswa->colormap = CORE(_apsw).colormap;
	_xswa->background_pixel = _xswa->border_pixel = 0;
	*_xvm |= CWColormap | CWBorderPixel | CWBackPixel;

	(*appPlusShellWidgetClass->core_class.superclass->core_class.realize)((Widget)_apsw, _xvm, _xswa);

	XSetWindowColormap(XtDisplay(_apsw), XtWindow(_apsw), CORE(_apsw).colormap);

#if 0 /* HMMM */
	XSetWMColormapWindows(XtDisplay(_apsw), XtWindow(_apsw), &XtWindow(_apsw), 1);
#endif /* 0 */

#if MOTIF & (XmREVISION >= 2)
	XmAddWMProtocols((Widget)_apsw, APPS(_apsw).saveYourself, APPS(_apsw).numberProtos);
#else
	(void)XSetWMProtocols(XtDisplay(_apsw), XtWindow(_apsw), APPS(_apsw).saveYourself, APPS(_apsw).numberProtos);
#endif /* MOTIF & (XmREVISION >= 2) */
}

#if NeedFunctionProtoTypes
static XtWidgetProc destroy(AppPlusShellWidget _apsw)
#else
static XtWidgetProc destroy(_apsw)
AppPlusShellWidget _apsw;
#endif /* NeedFunctionProtoTypes */
{
	XtRemoveAllCallbacks((Widget)_apsw, XtNsaveYourselfCallback);
}

#if NeedFunctionProtoTypes
static void WMProtoAction(AppPlusShellWidget _apsw, XEvent *_event, String *_parms, Cardinal *_numParms)
#else
static void WMProtoAction(_apsw, _event, _parms, _numParms)
AppPlusShellWidget _apsw;
XEvent *_event;
String *_parms;
Cardinal *_numParms;
#endif /* NeedFunctionProtoTypes */
{
	Atom proto;
	int i;

#if DEBUGGING
	printf("proto action\n");
#endif /* DEBUGGING */
 
	if (_event->type != ClientMessage)
		return;

	proto = (Atom)_event->xclient.data.l[0];

	for(i=0;i<APPS(_apsw).numberProtos;i++)
		if (APPS(_apsw).saveYourself[i] == proto)
		{
			XtCallCallbacks((Widget)_apsw, XtNsaveYourselfCallback, _event);
			break;
		}
}

#if NeedFunctionProtoTypes
static void toLower(char *_str1, char *_str2, int _length)
#else
static void toLower(_str1, _str2, _length)
char *_str1;
char *_str2;
int _length;
#endif /* NeedFunctionProtoTypes */
{
	int i;
	char *ptr;

	for(ptr=_str1,i=0;(ptr!=NULL) && (i<_length);ptr++,i++)
		*(_str2+i) = tolower(*ptr);
}

#if NeedFunctionProtoTypes
static Boolean cvtStringToVisualClass(Display *_display, XrmValuePtr _args,
	Cardinal *_numArgs, XrmValuePtr _from, XrmValuePtr _to, XtPointer *_data)
#else
static Boolean cvtStringToVisualClass(_display, _args, _numArgs, _from, _to, _data)
Display *_display;
XrmValuePtr _args;
Cardinal *_numArgs;
XrmValuePtr _from;
XrmValuePtr _to;
XtPointer *_data;
#endif /* NeedFunctionProtoTypes */
{
	char *lower;
	static int vClass;
	Boolean badConversion = FALSE;

	if (*_numArgs != 0)
	{
		XtAppWarningMsg(XtDisplayToApplicationContext(_display), "CvtStringToVisualClass", "wrongParamaters",
		"ResourceError",
		"cvtStringToVisualClass needs no arguments.",
		(String *)NULL, (Cardinal *)NULL);
	}

	lower = XtNewString(_from->addr);
	toLower(_from->addr, lower, strlen(_from->addr));

	vClass = (int)DEFAULT;

	if (!strncmp(lower, "staticgray", 10))
		vClass = StaticGray;
	else
	if (!strncmp(lower, "staticcolor", 11))
		vClass = StaticColor;
	else
	if (!strncmp(lower, "pseudocolor", 11))
		vClass = PseudoColor;
	else
	if (!strncmp(lower, "grayscale", 9))
		vClass = GrayScale;
	else
	if (!strncmp(lower, "truecolor", 9))
		vClass = TrueColor;
	else
	if (!strncmp(lower, "directcolor", 11))
		vClass = DirectColor;
	else
		badConversion = TRUE;

	XtFree(lower);

	if (badConversion)
		XtDisplayStringConversionWarning(_display, _from->addr, XtRVisualClass);
	else
	{
		if (_to->addr == NULL)
			_to->addr = (XtPointer)&vClass;
		else
		if (_to->size < sizeof(int))
			badConversion = TRUE;
		else
			*(int *)_to->addr = vClass;
		_to->size = sizeof(int);
	}

	return !badConversion;
}


#if NeedFunctionProtoTypes
static Boolean cvtStringToVisualID(Display *_display, XrmValuePtr _args,
		Cardinal *_numArgs, XrmValuePtr _from, XrmValuePtr _to, XtPointer *_data)
#else
static Boolean cvtStringToVisualID(_display, _args, _numArgs, _from, _to, _data)
Display *_display;
XrmValuePtr _args;
Cardinal *_numArgs;
XrmValuePtr _from;
XrmValuePtr _to;
XtPointer *_data;
#endif /* NeedFunctionProtoTypes */
{
		static VisualID value;
		Boolean badConversion;

		if (*_numArgs != 0)
		{
				XtAppWarningMsg(XtDisplayToApplicationContext(_display), "CvtStringToVisualID", "wrongParamaters",
				"ResourceError",
				"CvtStringToVisualID needs no arguments.",
				(String *)NULL, (Cardinal *)NULL);
		}

		badConversion = FALSE;

		value	= (VisualID)strtol(_from->addr, (char **)NULL, 0);

		if (_to->addr == NULL)
				_to->addr = (XtPointer)&value;
		else
		if (_to->size < sizeof(VisualID))
				badConversion = TRUE;
		else
				*(VisualID *)_to->addr = value;
				_to->size = sizeof(VisualID);

		return !badConversion;
}

#if HAS_XCC

#if NeedFunctionProtoTypes
XCC AppPlusShellGetXCC(Widget _w)
#else
XCC AppPlusShellGetXCC(_w)
Widget _w;
#endif /* NeedFunctionProtoTypes */
{
	AppPlusShellWidgetClass aswc;
	aswc = (AppPlusShellWidgetClass)appPlusShellWidgetClass;

	if (aswc->core_class.class_inited)
		return ACLASS(aswc).xcc;
	else
	{
		XtAppWarning(XtWidgetToApplicationContext(_w),
			"Attempting to access XCC when AppPlusShell is uninitialized!\n");
		return (XCC)NULL;
	}
}

#endif /* HAS_XCC */
